llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -3202,6 +3202,14 @@ Instruction *InstCombinerImpl::foldSelectOfBools(SelectInst &SI) {
     }
   }
 
+  // select (~b & a), a, b -> or a, b
+  // only for scalar types
+  if (match(CondVal, m_And(m_Not(m_Specific(FalseVal)), m_Specific(TrueVal))) &&
```

Comment:
Can `m_c_And` be used here? Also, you should avoid the transformation if `CondVal` has more than one use. I'd also suggest moving this code to around line 3088, to group it with other simplifications of select with conditions based on the two value operands.

Consider simplifying the mirror expresison `select (~a | b), a, b` to `~(a ^ b)`, if it isn't already.

Reply:
Good suggestions! I have included the pattern `select (~b | a), a, b`. The other two (`select (~a & b), a, b` and `select (~a | b), a, b`) are already simplified by the pass.